File: services/wisepen-chat-service/src/chat/api/endpoints/search_provider.py
from dependency_injector.wiring import Provide, inject
from fastapi import APIRouter, Depends

from chat.api.schemas.search_provider import (
    SearchProviderConfigResponse,
    SetCustomSearchProviderRequest,
    SetSearchProviderModeRequest,
)
from chat.application.api_service.search_provider import (
    SearchProviderConfigApiService,
    SearchProviderConfigView,
)
from chat.container import Container
from common.core.domain import R
from common.security import require_login

router = APIRouter()


@router.get("/getConfig", response_model=R[SearchProviderConfigResponse])
@inject
async def get_search_provider_config(
    user_id: str = Depends(require_login),
    service: SearchProviderConfigApiService = Depends(
        Provide[Container.search_provider_config_api_service]
    ),
):
    return R.success(data=_to_response(await service.get_config(user_id=user_id)))


@router.post("/setMode", response_model=R[SearchProviderConfigResponse])
@inject
async def set_search_provider_mode(
    req: SetSearchProviderModeRequest,
    user_id: str = Depends(require_login),
    service: SearchProviderConfigApiService = Depends(
        Provide[Container.search_provider_config_api_service]
    ),
):
    config = await service.set_mode(user_id=user_id, mode=req.mode)
    return R.success(data=_to_response(config))


@router.post("/setCustomProvider", response_model=R[SearchProviderConfigResponse])
@inject
async def set_custom_search_provider(
    req: SetCustomSearchProviderRequest,
    user_id: str = Depends(require_login),
    service: SearchProviderConfigApiService = Depends(
        Provide[Container.search_provider_config_api_service]
    ),
):
    config = await service.set_custom_provider(
        user_id=user_id,
        provider=req.provider,
        api_key=req.api_key,
    )
    return R.success(data=_to_response(config))


@router.post("/clearCustomProvider", response_model=R[SearchProviderConfigResponse])
@inject
async def clear_custom_search_provider(
    user_id: str = Depends(require_login),
    service: SearchProviderConfigApiService = Depends(
        Provide[Container.search_provider_config_api_service]
    ),
):
    config = await service.clear_custom_provider(user_id=user_id)
    return R.success(data=_to_response(config))


@router.post("/verifyProvider", response_model=R[SearchProviderConfigResponse])
@inject
async def verify_custom_search_provider(
    user_id: str = Depends(require_login),
    service: SearchProviderConfigApiService = Depends(
        Provide[Container.search_provider_config_api_service]
    ),
):
    config = await service.verify(user_id=user_id)
    return R.success(data=_to_response(config))


def _to_response(config: SearchProviderConfigView) -> SearchProviderConfigResponse:
    return SearchProviderConfigResponse(
        provider_mode=config.provider_mode,
        provider=config.provider,
        masked_key=config.masked_key,
        is_valid=config.is_valid,
    )

File: new/WisePenCloud-AI-clone/services/wisepen-chat-ui/src/api/searchProvider.ts
import { apiFetch, readApiData } from "./client";

export type SearchProviderMode = "default" | "custom";

export type SearchProviderName =
  | "serper"
  | "tavily"
  | "brave"
  | "serpapi"
  | "exa"
  | "perplexity"
  | "anysearch";

export type SearchProviderConfig = {
  mode: SearchProviderMode;
  provider: SearchProviderName | null;
  maskedKey: string | null;
  isValid: boolean;
};

type SearchProviderConfigDto = {
  provider_mode: SearchProviderMode;
  provider?: SearchProviderName | null;
  masked_key?: string | null;
  is_valid?: boolean;
};

export const searchProviders: Array<{
  value: SearchProviderName;
  label: string;
}> = [
    { value: "serper", label: "Serper" },
    { value: "tavily", label: "Tavily" },
    { value: "brave", label: "Brave" },
    { value: "serpapi", label: "SerpAPI" },
    { value: "exa", label: "Exa" },
    { value: "perplexity", label: "Perplexity" },
    { value: "anysearch", label: "AnySearch" },
  ];

export async function getSearchProviderConfig(): Promise<SearchProviderConfig> {
  const response = await apiFetch("/chat/searchProvider/getConfig");
  const data = await readApiData<SearchProviderConfigDto>(
    response,
    "加载搜索源配置失败",
  );
  return mapConfig(data);
}

export async function setSearchProviderMode(
  mode: SearchProviderMode,
): Promise<SearchProviderConfig> {
  const response = await apiFetch("/chat/searchProvider/setMode", {
    method: "POST",
    body: JSON.stringify({ mode }),
  });
  const data = await readApiData<SearchProviderConfigDto>(
    response,
    "设置搜索模式失败",
  );
  return mapConfig(data);
}

export async function setCustomSearchProvider(
  provider: SearchProviderName,
  apiKey: string,
): Promise<SearchProviderConfig> {
  const response = await apiFetch("/chat/searchProvider/setCustomProvider", {
    method: "POST",
    body: JSON.stringify({ provider, api_key: apiKey }),
  });
  const data = await readApiData<SearchProviderConfigDto>(
    response,
    "保存自定义搜索源失败",
  );
  return mapConfig(data);
}

export async function clearCustomSearchProvider(): Promise<SearchProviderConfig> {
  const response = await apiFetch("/chat/searchProvider/clearCustomProvider", {
    method: "POST",
  });
  const data = await readApiData<SearchProviderConfigDto>(
    response,
    "清除自定义搜索源失败",
  );
  return mapConfig(data);
}

export async function verifyCustomSearchProvider(): Promise<SearchProviderConfig> {
  const response = await apiFetch("/chat/searchProvider/verifyProvider", {
    method: "POST",
  });
  const data = await readApiData<SearchProviderConfigDto>(
    response,
    "验证自定义搜索源失败",
  );
  return mapConfig(data);
}

function mapConfig(dto: SearchProviderConfigDto): SearchProviderConfig {
  const mode = normalizeMode(dto.provider_mode);
  return {
    mode,
    provider: dto.provider ?? null,
    maskedKey: dto.masked_key ?? null,
    isValid: dto.is_valid ?? false,
  };
}

function normalizeMode(mode: SearchProviderConfigDto["provider_mode"]): SearchProviderMode {
  return String(mode || "default").toLowerCase() === "custom" ? "custom" : "default";
}
